parustiko/src/lib.rs

```diff
@@ -1,4 +1,65 @@
 mod errors;
 mod protocol;
+mod version_exchange;
 
-pub fn runner() {}
+use crypto::encryption::aes::{GenericArray, AES};
+use crypto::encryption::Encryption;
+use errors::VersionExchangeError;
+use protocol::DecodeRaw;
+use protocol::{BinaryProtocolPacket, Decode};
+use std::io::{Read, Write};
+use std::net::TcpStream;
+use std::time::Duration;
+use version_exchange::{KeyExchange, SshVersion};
+
+// Read bytes from stream until CR and LF ('\r\n') occur
+fn read_header<T: Read>(arr: &mut T) -> Result<Vec<u8>, VersionExchangeError> {
+    let mut header = Vec::with_capacity(51);
+    let mut prev_byte = None;
+
+    loop {
+        let mut byte = [0u8; 1];
```

Comment:
allocate this once and later just override

Reply:
can't we read directly into the vector, using a slice btw?

Reply:
I could replace this with `byte = 0u8` and then `arr.read_exact(std::slice::from_mut(&mut byte))` but I would like to stay with byte as a separate variable to compare with \r\n in an easy way (otherwise, if it was stored directly in vec, I would have to fetch the last element from vec everytime? - weird)

Reply:
what is weird about that?

Reply:
You can simply do something like this:
![image](https://github.com/user-attachments/assets/c89439e4-4dbf-4f38-bc5f-bf526ce1e6bd)

Reply:
You can simply do something like this:

![image](https://github.com/user-attachments/assets/c93be0f9-f4b7-48e2-9da4-0bf7ce4c5fe1)

Reply:
ok, if we limit the iterations it looks better, but e.g. if we have a vec with 10000000 elements, imo a new byte variable would be better than checking `arr[i-1] and arr[i]` every time.

Reply:
yeah, but in small arrays the lookup via index is practically O(1)